EIPS/eip-7951.md

```diff
@@ -0,0 +1,281 @@
+
+### Precompile
+
+We introduce `P256VERIFY` a precompile at the address `0x100` which performs ECDSA signature verification over the secp256r1 curve with a gas cost of `3450` gas
+
+### Curve Parameters
+
+The secp256r1 curve is fully defined by the following set of parameters:
+
+``` text
+Base field modulus = p = 0xffffffff00000001000000000000000000000000ffffffffffffffffffffffff
+Curve equation: y^2 = x^3 + ax + b (mod p)
+Curve coefficient a = 0xffffffff00000001000000000000000000000000fffffffffffffffffffffffc
+Curve coefficient b = 0x5ac635d8aa3a93e7b3ebbd55769886bc651d06b0cc53b0f63bce3c3e27d2604b
+Base point G:
+  Gx = 0x6b17d1f2e12c4247f8bce6e563a440f277037d812deb33a0f4a13945d898c296
+  Gy = 0x4fe342e2fe1a7f9b8ee7eb4a7c0f9e162bce33576b315ececbb6406837bf51f5
+Subgroup order = n = 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551
+Cofactor = h = 0x1
+```
+
+These parameters are standardized by NIST in SP 800-186[^1].
+
+### Fields and Groups
+
+The field Fp is defined as the finite field of size `p` with elements represented as integers between 0 and p-1 (both inclusive).
+
+The group G is defined as a set of Fp pairs (points) `(x,y)` such that either `(x,y)` is `(0,0)` (representing the point at infinity) or `x,y` satisfy the curve equation `y^2 = x^3 + ax + b (mod p)`.
+
+### Points and Encoding
+
+#### Field Elements Encoding
+
+A base field element (Fp) is encoded as `32` bytes by performing BigEndian encoding of the corresponding (unsigned) integer. The corresponding integer **must** be less than the field modulus `p`.
+
+#### Encoding of Points in G
+
+Points in G are encoded as byte concatenation of the respective encodings of the `x` and `y` coordinates. Total encoding length for a G point is thus `64` bytes.
+
+#### Point of Infinity Encoding
+
+For secp256r1, the point with coordinates `(0, 0)` (zeroes in Fp) is *not* on the curve, so a sequence of `64` zero bytes is used by convention to encode the point of infinity.
+
+#### Encoding of Scalars
+
+A scalar is encoded as `32` bytes by performing BigEndian encoding of the corresponding (unsigned) integer. The corresponding integer is **not** required to be less than or equal to the subgroup order `n`.
+
+#### Behavior on Invalid Inputs
+
+On inputs that cannot be valid encodings of field elements or points, the precompile *must* return `` (failure).
+
+### ABI for `P256VERIFY` Operation
+
+#### Input
+
+`P256VERIFY` call expects `160` bytes as input that is interpreted as byte concatenation of:
+
+- `32` bytes: message hash `h`
+- `32` bytes: signature component `r`
+- `32` bytes: signature component `s`
+- `32` bytes: public key x-coordinate `qx`
+- `32` bytes: public key y-coordinate `qy`
+
+#### Output
+
+Output is `32` bytes on successful verification and `0` bytes on failure:
+
+- `0x0000000000000000000000000000000000000000000000000000000000000001` for valid signatures
+- `` for invalid signatures or invalid inputs
```

Comment:
This is known and documented to be bad. Why keep it that way?

Reply:
for reference: https://ethereum-magicians.org/t/eip-7212-precompiled-for-secp256r1-curve-support/14789/96

Reply:
This forces us to do [that](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/P256.sol#L84-L111), which is terrible.

Reply:
I mean the vast majority of this is just checking whether the precompile even exists.

Reply:
@Amxx @CarlBeek here's the full backstory for how it ended up this way.

https://ethereum-magicians.org/t/eip-7212-precompiled-for-secp256r1-curve-support/14789/96